983 improve mobile support

juzen2003 opened the pull request:
- Fixes #983 
- Fixes #1005 
- Were any Django, import pipeline, table_schema, or dictionary files modified? N
  - Database used: opus3_test_200211
  - All Django tests pass: Y
- Were any JavaScript or CSS files modified? Y
  - JSHINT run on all affected files: Y 
  - Tested on Chrome, Firefox, Safari, and iOS: Y
    (Remember to test all browser sizes)
  - Tested for race conditions (with delayed API calls if applicable): NA

Description of changes:
General: 
- Change supported browser size to 550px width & 270px height
- When browser width is less than 580px OR height is shorter than 400px:
    - Shrink the modal header 
    - Hide the footer
    - Set body font-size to be 75%
    - Set reset buttons font-size to be 90% (in search tab)
    - Shrink the height of main nav
- When browser width is less than 580px AND height is shorter than 400px:
    - Set the table header font size to 75% to have a better table display

Search:
- Recalculate mainNavHeight in searchBarContainerHeight & searchWidgetHeightChanged to have the correct height of search menu sidebar and widgets area.
- When browser width is less than 580px OR height is shorter than 630px:
    - Hide the feedback button
- When browser width is less than 580px OR height is shorter than 400px:
    - Reduce the padding around the whole search tab
    - Expand the width of widgets area

Browse:
- When browser height is shorter than 400px:
    - Hide download csv button in select metadata menu
- Update the logic in calculateGalleryHeight, take overlapped nav into consideration. This will remove the little gap between the end of gallery contents and the footer.

Cart:
- When browser height is shorter than 400px:
    - Hide product type select/deselect buttons in download panel
    - Hide "Download Options" in download panel
- When browser width is less than 580px OR height is shorter than 630px:
    - Hide download zip buttons in download panel
- When browser width is less than 580px OR height is shorter than 400px:
    - Reduce the top & bottom spaces around "Download Options" 
    - Reduce the space around data in download panel card body
    - Reduce the space around "Download Data" in download panel 

Detail:
- When browser height is less than 400px:
    - Keep the detail image stay at the right of the page
- When browser width is less than 768px AND height is shorter than 400px:
    - Make observation detail area shrinkable
    - Make observation detail area take up 70% of width and image take up 30%
    - Reduce the spaces around detail tab

Metadata modal for each observation:
- Make sure image always stays at the left in metadata modal
- Make sure metadata modal is not covered by main nav 
- Make sure hamburger menu is not covered by the modal

Known problems:
Mobile touch over-scrolling:
- In Pixel with Chrome (Firefox doesn't have this issue), in browse tab, there will be a huge gap at the end of gallery contents and footer.
- In iOS or Android devices with smaller screen (like iphone 8 or iphone se), in browse/cart tab, the 2nd navbar & gallery area can be over-scrolled. (The issue is not happened on iphone pro max).

juzen2003 committed:
Update browser size dialog messages to ask users to rotate their mobile
device to landscape mode for wider screen. Add the styling to prevent
x-scrolling in search tab.
- In opus.css (line 158-159), add styling to make sure x-scrolling is
  disabled in search tab.
- In opus.js (line 1275-1278), update browser size dialog messages.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 983_improve_mobile_support

juzen2003 committed:
Hide app footer when screen width is less than 600px. Update the calculation
of mainNavHeight in searchBarContainerHeight & searchWidgetHeightChanged.
- In opus.css (line 2769-2774), hide app footer when screen width is less
  than 600px;
- In search.js (line 1069-1070, 1037-1038), update the calculation of
  mainNavHeight in searchBarContainerHeight & searchWidgetHeightChanged.
  This will make sure when screen width is narrow, and navbar is expanded,
  we can still get the correct mainNavHeight.

juzen2003 committed:
Make OPUS search tab work in iphone SE (Search tab is fine tuned).
- Set the browserSupport min width to 550px, and min height to 300px.
  (opus.js, line 122-123)
- Shrink the font size to 80% and reset buttons font size to 90%.
  (opus.css, line 2776-2782)
- Hide the feedback buttons when screen is too small for the feedback
  form to be fully displayed. (opus.css, line 2785-2790)

juzen2003 committed:
Reduce redundant spaces in search tab when browser width is less than 575px.
- Shrink the height of main nav (opus.css, line 2789-2793).
- Reduce the space around search sidebar & widgets when screen narrow
  (opus.css, line 2795-2801).
- Expand the width of widgets area when screen is narrow (opus.css, line
  2803-2807).

juzen2003 committed:
Shrink the modal header when screen is short or narrow. Hide download csv
in select metadata menu when screen is short.
- Shrink the modal header when screen is short or narrow (opus.css, line
  2817-2822).
- Hide download csv button in select metadata menu when screen is short
  (opus.css, line 2824-2828).

juzen2003 committed:
Update calculateGalleryHeight to take overlapped navbar height into
consideration when calculating the gallery contents height. This will
remove the little white gap at the end of the gallery contents and the
footer.
- Update the logic in calculateGalleryHeight to remove the little white
  gap at the end of the gallery contents and the footer (browse.js, line
  2055-2058)

juzen2003 committed:
Fine tune download panel in cart tab.
- Center "Download Data" in download panel header (base.html, line 244,
  opus.css, line 1607-1612).
- Change the supported browser size to width 550, height 270 (opus.js,
  line 123).
- Fine tune the download panel when screen is narrow and short (opus.css,
  line 2833-2883).

juzen2003 committed:
Tune the detail page to support mobile device.
- Update adjustDetailHeight. When detail image is moved to the top of
  detail left pane, we have to account for the height of detail image as
  well when calculating the containerHeight (detail.js, line 130)
- When browser height is less than 400px:
    - Keep the detail image stay at the right of the page (opus.css,
      line 2897-2903)
- When browser width is less than 768px AND height is shorter than 400px:
    - Make observation detail area shrinkable (opus.css, line 2908-2911)
    - Make observation detail area take up 70% of width and image take up
      30% (opus.css, line 2913-2917)
    - Reduce the spaces around detail tab (opus.css, line 2919-2925)

juzen2003 committed:
Tune the metadata modal to support mobile devices.
- Make sure image always stays at the left in metadata modal (opus.css,
  line 2908-2911)
- Make sure metadata modal is not covered by main nav (opus.css, line
  1507-1509)

juzen2003 committed:
Remove unwanted const at detail.js (detail.js, line 9)

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 983_improve_mobile_support

juzen2003 committed:
Make sure hamburger menu will not be covered by metadata modal (opus.css,
line 2674-2677)

juzen2003 committed:
Make "+ (OR)" have sans-serif font (opus.css, line 2617-2623).

juzen2003 committed:
Update responsiveness based on Rob's github feedback.
- The 2nd nav in browse tab:
    - When screen width is less than 580 AND height is less than 400px:
        - Remove "Download CSV (all results)" from the nav bar and move
          the Observation # slider up
- Confirm modal:
    - When screen width is less than 580 OR height is less than 400px:
        - Override .modal-header padding to top/bottom 0.25rem &
          left/right to 0.5rem
        - Override .modal-body padding to top/bottom 0.25rem &
          left/right to 0.5rem
        - Override .modal-footer padding to top/bottom 0.25rem &
          left/right to 0.5rem
- Select metadata menu:
    - When screen height is less than 400px:
        - Increase the height of the menu by removing the 55px gap between
          the modal & bottom edge of the screen
- Cart download panel:
    - When screen width is less than 1100px OR height is less than 460px:
        - Cart download options pane will be hidden, and "Download Data"
          button will show up
    - When screen height is less than 400px:
        - Move "Download Data" to the same as cart actions buttons
        - Remove the left pointing arrow
    - When screen width is less than 580px AND height is less than 400px:
        - Reduce spaces to make sure cart actions buttons are wrapped into
          the next line when screen is narrow & short
    - When screen height is less than 500px:
        - Hide select/deselect product types button
- Thumbnail hamburger menu:
    - When screen height is less than 400px:
        - Reduce the spaces of each menu item in thumbnail hamburger
- Help menu:
    - When screen height is less than 400px:
        - Reduce the spaces of each menu item
        - Reduce the font size to 90%
- Detail tab:
    - Make sure the text below the image are the same size as ones in
      Observation detail

juzen2003 committed:
Update responsiveness based on Debby's feedback. Here are details:
- Main nav:
    - When screen width is less than 580px OR height is less than 400px:
        - Reduce the padding around selected tab
        - Reduce the hamburger icon
- Widget:
    - When screen width is less than 580px OR height is less than 400px:
        - Reduce the padding around widget header
        - Reduce the spaces between each widget
- Slider:
    - When screen width is less than 580 OR height is less than 400px:
        - Reduce the slider font size to 90%
- Modal:
    - When screen width is less than 580 OR height is less than 400px:
        - Reduce the modal buttons size
    - When screen width is less than 580 OR height is less than 400px:
        - Keep the spaces for detail metadata modal in browse & cart tab
          to have the pretty display
- Detail:
    - When screen height is less than 400px:
        - Reduce the spaces around "Observation Detail"
        - Reduce the spaces around "OPUS ID"

juzen2003 committed:
Reduce the size of font awesome icon in slide show when screen width is
less than 580 OR height is less than 400px.

rfrenchseti reviewed (changes requested):
Search Tab:

- As you narrow the browser, the Reset buttons don't stay centered under the horizontal line above them and there are a couple of strange twitches in horizontal position which seem like they could be smoothed out.
- The same twitch at browser width 700 and 705 cause word wrapping under Surface Geometry Constraints to flip back and forth. It seems like there should only be one break point instead of two.
- When the font changes size at 580, the widget titles don't change.
- Do you think it would be worth changing the font size somewhere larger, like 700, instead of 580? Below around 700 the search tab gets awfully crunched.
- This may be hard to do, but as the width drops below 970, the min/max fields go to two lines and the input fields fill the complete width of the widget. But the +OR or trashcan wrap to the next line. It seems like the input fields could be kept narrowed and the +OR/trashcan kept next to the second line longer. Something similar happens for string input fields. The qtype and +OR wrap long before it seems like they need to. The input field could be made narrower first.

Cart Tab:
- What is the reason for removing the product pane when the height is < 460?

Help Menu:
- We should remove "Contact Us" from the Help menu at the same time we remove the Feedback tab due to size.

Detail Tab:
- When the font size shrinks, I don't think the various category headers change size.

- How hard would it be to shrink the thumbnail size whenever we shrink the font size?

juzen2003 committed:
Update responsiveness behaviors. Here are details:
General:
- Change the breakpoint of shrinking font-size from 580 to 700px
Search tab:
- In browser resize event, call the non-debounced version of
  o_search.searchHeightChanged() to prevent the horizontal position twitch
  caused by the delayed calculation of search sidebar & widget height.
- When screen width is less than 700px OR height is less than 400px
    - When font shrinks, remove the top margin to make reset buttons
      horizontally centered
    - Reduce widget title font-size to 75%

juzen2003 committed:
Update the responsiveness on cart download pane & slide panel. Here are
details:
General:
    - When screen width is less than 700px OR height is less than 630px:
        - Hide "Contact Us" in help panel
Cart tab:
    - Update the logic of calculating product type table height in
      adjustProductInfoHeight
    - Keep the download data left pane when screen is short. Only hide
      it when screen width is less than 1100px
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the size of "Metadata CSV" button
        - Reduce the space around "Download Data" in download pane

juzen2003 committed:
Convert mobile support padding/margin units from px/rem to em.

juzen2003 committed:
Update stylings for responsiveness, and here are details:
Detail tab:
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the font-size of category headers in observation detail
Mini thumbnail in table view:
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the size of zoom-in mini-thumbnail on hover
Select metadata menu:
    - When screen width is less than 700px:
        - Position "Download CSV (all results)" button right above the
          modal footer when screen is narrow
Metadata modal:
    - Add some gap between font awesome icons and metadata info area

juzen2003 committed:
Update the stylings for responsiveness. Here are details:
Gallery thumbnail:
- When screen width is less than 700px OR height is less than 400px:
    - Reduce the thumbnail container from 100x100 to 90x90 (10% off)
      when font size is changed
STRING & RANGE input:
- Set the width of input.STRING & input.RANGE to make sure it can be
  shrunk more when screen is shrinking (before reaching to the breakpoint
  at 940px)

rfrenchseti reviewed (changes requested):
- The Feedback tab / contact us disappears at 700px width. It doesn't need to go away until 560px which is what I thought the previous behavior was.
- In the cart download pane, the "download archive" buttons go away at height 630px, and the "download CSV' button is still here. I don't see any reason for the archive/URL buttons to go away. They also go away at height 700px. Frankly I don't see any reason the buttons ever need to go away.
- In the slideshow, I think there's too much vertical space between the bottom icons and the bottom of the metadata container.
- At width 1100px, there's a weird change in the width of the string and range input fields. I don't think the change needs to happen there because there is still plenty of whitespace to the right of the fields. The browser could be narrower before shrinking the inputs.
- I don't think we were talking about the same "twitch" around 700. Whatever it was, it's gone now though.

rfrenchseti reviewed (changes requested):
- On the iPhone, in table view, click the "add all to cart" icon. Most of the time the modal verification dialog will show up at a weird z-layer behind the table header. Sometimes it won't show up at all.
- On the iPhone, in table view, click the hamburger then "start range". The green square will appear but won't be animated. It seems to work fine on the gallery view, or if you start the range on the gallery and then switch to the table. Sounds like it might be a browser bug.

juzen2003 committed:
Update the styling for responsiveness, and here are details:
- Feedback & contact us:
    - When screen width is less than 560px OR height is less than 630px:
        - Hide feedback tab & contact us option in help menu
- Data Archive & URL Archive buttons in download pane:
    - Both buttons will not go away now
    - When screen width is less than 700px OR height is less than 400px:
        - Reduce the spaces around both buttons
- Metadata modal:
    - Reduce the gap between the bottom icons and the container of
      metadata info

juzen2003 committed:
Update styling for mobile support. Here are details:
- Input.RANGE:
    - Remove the width styling at 1100px breakpoint
- "Add All Results to Cart" table header icon:
    - Set title attributes to the button tag that wraps the cart icon
      (i tag). This will make sure default tooltip is properly displayed
      on PC, and mobile device touch start will trigger the button click
      event handler instead of the icon touchstart event handler

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -259,7 +260,8 @@ Not sure if this is the desired behavior
 to 100%, this will make the shrinking look smoother */
-@media (max-width: 970px) {
+@media (max-width: 940px) {
+/* @media (max-width: 970px) { */
```
Remove comment

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -1373,7 +1373,7 @@ to stay in one line */
     /* Move the last item in flexbox to the 2nd line */
     #browse .browse-nav-container .navbar-nav {
-        flex-wrap: wrap;
+        /* flex-wrap: wrap; */
```
Remove comment

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -1577,6 +1562,7 @@ to stay in one line */
 
 /* Remove the nav-link for opening up download data panel when screen
 is wide enough */
+/* @media (min-width: 1101px) and (min-height: 461px) { */
```
Remove comment

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -2764,3 +2757,366 @@ when wrapping into the 2nd line */
 }
+
+/* The followings are for mobile support (width: 550px and height: 270px) */
+@media (max-width: 700px), (max-height: 400px){
```
Add space before {

rfrenchseti reviewed (changes requested):
- When the browser is short enough that the footer goes away, and then you start add/remove range, the text bubble that pops up in the lower left is now too tall. I think it's assuming that the footer is still there.
- In table view, when you start add/remove range and the green square is pulsing, if you scroll the table up that particular line entry will actually move over the table headers. It has the wrong Z value somehow. This happens on all browsers.
- I still don't see the square pulsing in the table view on the iPhone. I also don't see the changes you made in the actual code. Did you forget a push?

juzen2003 committed:
Update the styling for responsiveness, and here are details:
- Range select info box at lower left corner:
    - When screen width is less than 700px OR height is less than 400px:
        - When footer is gone, make sure the range select info box has
          the proper height
- hvr-ripple-in:
    - Set z-index to auto to avoid overlapping with table header

rfrenchseti reviewed (changes requested):
There is still one really fatal problem with small browsers. It now doesn't count the number of thumbnails properly. If you set your browser to the size of an iPhone 7 (which has 7 thumbnails across), the slider will only increment by 6. This makes a mess of everything else.

juzen2003 committed:
Update o_browse.imageSize to 90 when screen width is less than 700px or
height is less than 400px. This will make sure slider number gets calculated
correctly.
- Add comments in opus.css to make sure the breakpoints for mobile support
  will match the constantbrowserThresholdWidth & browserThresholdHeight
  in opus.js. (opus.css, line 2759-2764)
- Add variables to store threshold width & height when font & thumbnail
  size start to shirnk. (opus.js, line 137-140)
- Add function updateImageSize to update o_browse.imageSize. It will get
  called in adjustBrowseHeight when browser resize happened. This will
  make sure the calculation for gallery image numbers per row is correct.

juzen2003 committed:
Add comments in opus.css for .op-thumbnail-container min-width & min-height.

dstopp reviewed (approved):
looks good; note that merge of 362 will likely have a lot of conflicts..

juzen2003 committed:
Reduce the space around "Download Data" in download panel when screen is
narrow or short (cart tab)

juzen2003 committed:
Reduce the font size & padding around the card header for download panel
& help panel.
- Reduce the space around "Download Data" in download panel & the header
  text in help panel when screen is narrow or short.
- Shrink the font size of "Download Data" in download panel & the header
  text in help panel to 90% when screen is narrow or short.
- Shrink "x" icon in download & help panel when screen is narrow or short.

juzen2003 committed:
Update to reduce the font size of the header text in download & help panel
to 75%.
